pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/ReplicaGroupSegmentAssignmentStrategy.java

```diff
@@ -0,0 +1,270 @@
+ *   if there is a tie. Repeat the process for all the partitions in one replica, and mirror the assignment to other
+ *   replicas. With this greedy algorithm, the result is deterministic and with minimum segment moves.
+ */
+public class ReplicaGroupSegmentAssignmentStrategy implements SegmentAssignmentStrategy {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ReplicaGroupSegmentAssignmentStrategy.class);
+
+  private ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private String _tableNameWithType;
+  private String _partitionColumn;
+
+  @Override
+  public void init(ZkHelixPropertyStore<ZNRecord> propertyStore, TableConfig tableConfig) {
+    _propertyStore = propertyStore;
+    _tableNameWithType = tableConfig.getTableName();
+    ReplicaGroupStrategyConfig strategyConfig = tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
+    _partitionColumn = strategyConfig != null ? strategyConfig.getPartitionColumn() : null;
+
+    if (_partitionColumn == null) {
+      LOGGER.info("Initialized ReplicaGroupSegmentAssignmentStrategy for table: {} without partition column",
+          _tableNameWithType);
+    } else {
+      LOGGER.info("Initialized ReplicaGroupSegmentAssignmentStrategy for table: {} with partition column: {}",
+          _tableNameWithType, _partitionColumn);
+    }
+  }
+
+  @Override
+  public List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
+      InstancePartitions instancePartitions) {
+    // Fetch partition id from segment ZK metadata if partition column is configured
+    int partitionId = 0;
+    if (_partitionColumn == null) {
+      Preconditions.checkState(instancePartitions.getNumPartitions() == 1,
+          "The instance partitions should contain only 1 partition");
+    } else {
+      OfflineSegmentZKMetadata segmentZKMetadata =
+          ZKMetadataProvider.getOfflineSegmentZKMetadata(_propertyStore, _tableNameWithType, segmentName);
+      Preconditions
+          .checkState(segmentZKMetadata != null, "Failed to fetch segment ZK metadata for table: %s, segment: %s",
+              _tableNameWithType, segmentName);
+      partitionId = getPartitionId(segmentZKMetadata, instancePartitions);
+    }
+
+    // First assign the segment to replica 0
+    List<String> instances = instancePartitions.getInstances(partitionId, 0);
+    int[] numSegmentsAssigned = SegmentAssignmentUtils.getNumSegmentsAssigned(currentAssignment, instances);
+    int minNumSegmentsAssigned = numSegmentsAssigned[0];
+    int instanceIdWithLeastSegmentsAssigned = 0;
+    int numInstances = numSegmentsAssigned.length;
+    for (int instanceId = 1; instanceId < numInstances; instanceId++) {
+      if (numSegmentsAssigned[instanceId] < minNumSegmentsAssigned) {
+        minNumSegmentsAssigned = numSegmentsAssigned[instanceId];
+        instanceIdWithLeastSegmentsAssigned = instanceId;
+      }
+    }
+
+    // Mirror the assignment to all replicas
```

Comment:
Are we planning to remove config for `mirror` and make it by default? +1 for removing config

The downside of this will be that if we ever want to support different # of servers for each replica group, we will not be able to do mirroring anymore. This code will add the constraint on that scenario while no-mirroring will naturally support that.

Currently, I cannot think of any use case that needs the relaxation. You can also double check on this.

Reply:
Yes. We are going to always use mirror. I cannot think of any use case which needs different number of servers for each replica group, and we don't support that as well. If we actually need that, we can add the support later.